### PR TITLE
Adjust delete key and add a API to put sth

### PR DESCRIPTION
**Pull request by xiekeyi98**
### What problem does this PR solve?

Add API to control etcd without etcd binary.
Because:
1. Sometimes the node does not delete successfully, we should delete it manually.
2. To test the DDL schema, we need to put some fake key to froze schema.

### What is changed and how it works?

Use etcd api_gateway to control etcd.

### Check List

Tests
- Unit test

### Side effects

The API may be abused.

### CODE REVIEW

**Comment by xiekeyi98** [comment]
test

**Comment by xiekeyi98** [comment]
When we add a schema version key(fake) , then execute DML operation.
image

I found it will took double lease time to be done.

**Comment by xiekeyi98** [comment]
@winkyao  
@zimulala 
PTAL.

File: cmd/etcd.go [hunk 1]
```diff
@@ -61,22 +61,22 @@ func newShowDDLInfoCommand() *cobra.Command {
 	return m
 }
 
-// newDelOwnerCampaign return a delete owner campaign subcommand of EtcdCommand
-func newDelOwnerCampaign() *cobra.Command {
+// newDelKeyCommand return a delete key subcommand of EtcdCommand
```
Comment (zimulala) [inline cmd/etcd.go]: s/return/returns
Add `.` at the end of a line.

File: cmd/etcd.go [hunk 2]
```diff
@@ -61,22 +61,22 @@ func newShowDDLInfoCommand() *cobra.Command {
...
+		Use:   "delkey",
+		Short: "del key by `del key [key]`",
+		Run:   delKeyCommandFunc,
 	}
 	return m
 }
 
-// newDelSchemaVersion return a delete schema version subcommand of EtcdCommand
-func newDelSchemaVersion() *cobra.Command {
+// newPutKeyCommand return a put key subcommand of EtcdCommand
```
Comment (zimulala) [inline cmd/etcd.go]: Ditto.

File: cmd/etcd.go [hunk 3]
```diff
@@ -113,76 +113,79 @@ func showDDLInfoCommandFunc(cmd *cobra.Command, args []string) {
...
 	if len(args) != 1 {
-		cmd.Println(cmd.UsageString())
+		cmd.Printf("Only one argument!")
 		return
 	}
 
-	leaseID := args[0]
+	key := args[0]
+
+	if !strings.HasPrefix(key, "/tidb/ddl/") {
```
Comment (zimulala) [inline cmd/etcd.go]: Could we define `/tidb/ddl/` as a constant?

File: cmd/etcd_test.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
+package cmd
```
Comment (zimulala) [inline cmd/etcd_test.go:14]: Please add a license.

File: cmd/etcd.go (hunk 1 above)
Comment (xiekeyi98) [reply cmd/etcd.go]: done.

File: cmd/etcd.go (hunk 2 above)
Comment (xiekeyi98) [reply cmd/etcd.go]: done.

File: cmd/etcd_test.go (hunk 4 above)
Comment (xiekeyi98) [reply cmd/etcd_test.go:14]: done.

File: cmd/etcd.go (hunk 3 above)
Comment (xiekeyi98) [reply cmd/etcd.go]: Done.
Do we need to make its scope global?
( I think in function is enough.

**Comment by winkyao** [comment]
@xiekeyi98 Please don't leave the description empty.

**Comment by xiekeyi98** [comment]
> @xiekeyi98 Please don't leave the description empty.

Done.
Sorry for my fault.
Because this repo has not PR template, I would have thought it does not need it.

File: cmd/etcd.go [hunk 5]
```diff
@@ -113,76 +115,79 @@ func showDDLInfoCommandFunc(cmd *cobra.Command, args []string) {
...
 }
 
-func delOwnerCampaign(cmd *cobra.Command, args []string) {
+func delKeyCommandFunc(cmd *cobra.Command, args []string) {
 	if len(args) != 1 {
-		cmd.Println(cmd.UsageString())
+		cmd.Printf("Only one argument!")
+		return
+	}
+	apiScope := "/tidb/ddl"
```
Comment (winkyao) [inline cmd/etcd.go]: Allow to delete key range in `/tidb/ddl` is dangarous.

Comment (xiekeyi98) [reply cmd/etcd.go]: Do we need to delete key in `/tidb/ddl/fg/owner` ?

**Comment by xiekeyi98** [comment]
I added check key exist.

image

File: cmd/etcd.go [hunk 6]
```diff
@@ -61,134 +61,170 @@ func newShowDDLInfoCommand() *cobra.Command {
...
+func delKeyCommandFunc(cmd *cobra.Command, args []string) {
+	if len(args) != 1 {
+		cmd.Printf("Only one argument!")
 		return
 	}
-	res, err := dail(req)
-	if err != nil {
-		cmd.Printf("Failed to show DDLInfo: %v\n", err)
+	var apiScope [2]string
+	apiScope[0] = "/tidb/ddl/fg/owner/"
```
Comment (winkyao) [inline cmd/etcd.go]:
```
ddlOwnerKeyPrefix := "/tidb/ddl/fg/owner/"
ddlAllSchemaVersionsPrefix = "/tidb/ddl/all_schema_versions/"

if !(strings.HasPrefix(key, ddlOwnerKeyPrefix) || strings.HasPrefix(key, ddlAllSchemaVersionsPrefix))
```

Comment (xiekeyi98) [reply cmd/etcd.go]: Done.

File: cmd/etcd.go [hunk 7]
```diff
@@ -61,134 +61,171 @@ func newShowDDLInfoCommand() *cobra.Command {
...
 	}
-	res, err := dail(req)
-	if err != nil {
-		cmd.Printf("Failed to show DDLInfo: %v\n", err)
+
+	key := args[0]
+	ddlOwnerKeyPrefix := "/tidb/ddl/fg/owner/"
+	ddlAllSchemaVersionsPrefix := "/tidb/ddl/all_schema_versions/"
+	if !(strings.HasPrefix(key, ddlOwnerKeyPrefix) || strings.HasPrefix(key, ddlAllSchemaVersionsPrefix)) {
+		cmd.Printf("This function only for delete something about DDL")
```
Comment (zimulala) [inline cmd/etcd.go]: Do we need add `\n` at the end of this log? We can print the `args` information.

File: cmd/etcd_test.go [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+	_, _, err = executeCommandC(cmd, args...)
+	c.Assert(err, IsNil)
+}
+
+func (s *etcdTestSuite) TestDelKey(c *C) {
+	testKey := "test"
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusOK)
+		c.Assert(r.Method, Equals, "POST")
+		c.Assert(r.URL.EscapedPath(), Equals, "/v3/kv/deleterange")
```
Comment (zimulala) [inline cmd/etcd_test.go]: Using `rangeDelPrefix ` is better.

File: cmd/etcd_test.go [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+	c.Assert(output, NotNil)
+}
+
+func (s *etcdTestSuite) TestPutKey(c *C) {
+	testKey := "test"
+	testValue := "test"
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusOK)
+		c.Assert(r.Method, Equals, "POST")
+		c.Assert(r.URL.EscapedPath(), Equals, "/v3/kv/put")
```
Comment (zimulala) [inline cmd/etcd_test.go]: Ditto.

File: cmd/etcd.go [hunk 10]
```diff
@@ -61,134 +61,171 @@ func newShowDDLInfoCommand() *cobra.Command {
...
+		cmd.Printf("Only one argument!")
 		return
 	}
-	res, err := dail(req)
-	if err != nil {
-		cmd.Printf("Failed to show DDLInfo: %v\n", err)
+
+	key := args[0]
+	ddlOwnerKeyPrefix := "/tidb/ddl/fg/owner/"
+	ddlAllSchemaVersionsPrefix := "/tidb/ddl/all_schema_versions/"
```
Comment (zimulala) [inline cmd/etcd.go]: `/tidb/ddl/all_schema_versions` is used in multiple places. I think to define it as a global constant is better.

File: cmd/etcd.go [hunk 11]
```diff
@@ -61,134 +61,171 @@ func newShowDDLInfoCommand() *cobra.Command {
...
+			if kk == "key" && vv == key {
+				findKey = true
+				break
+			}
+		}
+	}
 
-	leaseID := args[0]
+	if !findKey {
+		cmd.Printf("Failed to delete key: Key not found!")
```
Comment (zimulala) [inline cmd/etcd.go]: s/`Printf`/`Println` ?

File: cmd/etcd.go [hunk 12]
```diff
@@ -61,134 +61,171 @@ func newShowDDLInfoCommand() *cobra.Command {
...
 	}
 
-	cmd.Println(res)
-}
-
-func delOwnerCampaign(cmd *cobra.Command, args []string) {
-	if len(args) != 1 {
-		cmd.Println(cmd.UsageString())
+	findKey := false
+	var jsn struct {
```
Comment (zimulala) [inline cmd/etcd.go]: It's used in multiple places. Could we define it as a global variable?

File: cmd/etcd.go (hunk 7 above)
Comment (xiekeyi98) [reply cmd/etcd.go]: I think print `args` is not necessary.
Because this error is usually close to what you input.

File: cmd/etcd.go (hunk 12 above)
Comment (xiekeyi98) [reply cmd/etcd.go]: Now I change it , and it will be used only once.

File: cmd/etcd.go (hunk 11 above)
Comment (xiekeyi98) [reply cmd/etcd.go]: Done.

File: cmd/etcd.go (hunk 10 above)
Comment (xiekeyi98) [reply cmd/etcd.go]: Done.

File: cmd/etcd.go [hunk 13]
```diff
@@ -61,138 +62,137 @@ func newShowDDLInfoCommand() *cobra.Command {
...
-// newDelSchemaVersion return a delete schema version subcommand of EtcdCommand
-func newDelSchemaVersion() *cobra.Command {
+// newPutKeyCommand returns a put key subcommand of EtcdCommand.
+func newPutKeyCommand() *cobra.Command {
 	m := &cobra.Command{
-		Use:   "delschema",
-		Short: "delete schema version by DDLID",
-		Run:   delSchemaVersion,
+		Use:   "putkey",
+		Short: "[ONLY FOR TEST!] put key in schema version by `putkey [key] [value]`",
```
Comment (zimulala) [inline cmd/etcd.go]:
```suggestion
		Short: "[ONLY FOR TEST!] put a key in the path of TiDB schema versions by `putkey [key] [value]`",

File: cmd/etcd.go [hunk 14]
```diff
@@ -61,138 +62,137 @@ func newShowDDLInfoCommand() *cobra.Command {
...
-// newDelOwnerCampaign return a delete owner campaign subcommand of EtcdCommand
-func newDelOwnerCampaign() *cobra.Command {
+// newDelKeyCommand returns a delete key subcommand of EtcdCommand.
+func newDelKeyCommand() *cobra.Command {
 	m := &cobra.Command{
-		Use:   "delowner [LeaseID]",
-		Short: "delete DDL Owner Campaign by LeaseID",
-		Run:   delOwnerCampaign,
+		Use:   "delkey",
+		Short: "del key by `delkey [key]`",
```
Comment (zimulala) [inline cmd/etcd.go]:
```suggestion
		Short: "Delete the key associated with DDL by `delkey [key]`",

File: cmd/etcd.go [hunk 15]
```diff
@@ -61,138 +62,137 @@ func newShowDDLInfoCommand() *cobra.Command {
...
+func delKeyCommandFunc(cmd *cobra.Command, args []string) {
+	if len(args) != 1 {
+		cmd.Println("Only one argument!")
+		return
+	}
+
+	key := args[0]
+	ddlOwnerKeyPrefix := "/tidb/ddl/fg/owner/"
+	if !(strings.HasPrefix(key, ddlOwnerKeyPrefix) || strings.HasPrefix(key, ddlAllSchemaVersionsPrefix)) {
+		cmd.Println("This function only for delete something about DDL")
```
Comment (zimulala) [inline cmd/etcd.go]: s/something/the key-value

**Review by zimulala — COMMENTED** [review]
LGTM

File: cmd/etcd.go [hunk 16]
```diff
@@ -61,138 +62,137 @@ func newShowDDLInfoCommand() *cobra.Command {
...
+
+func delKeyCommandFunc(cmd *cobra.Command, args []string) {
+	if len(args) != 1 {
+		cmd.Println("Only one argument!")
+		return
+	}
+
+	key := args[0]
+	ddlOwnerKeyPrefix := "/tidb/ddl/fg/owner/"
+	if !(strings.HasPrefix(key, ddlOwnerKeyPrefix) || strings.HasPrefix(key, ddlAllSchemaVersionsPrefix)) {
```
Comment (winkyao) [inline cmd/etcd.go:102]: You need to check the key must not equals to "/tidb/ddl/fg/owner/" or ddlAllSchemaVersionsPrefix

File: cmd/etcd.go [hunk 17]
```diff
@@ -61,138 +62,137 @@ func newShowDDLInfoCommand() *cobra.Command {
...
-
-	leaseID := args[0]
-
-	var para = &parameter{
-		Key: base64Encode(delOwnerCampaignPrefix + leaseID),
+	var putParameter struct {
+		Key   string `json:"key"`
+		Value string `json:"value"`
 	}
+	putParameter.Key = base64Encode(ddlAllSchemaVersionsPrefix + args[0])
```
Comment (winkyao) [inline cmd/etcd.go:148]: check if len(args[0]) > 0

File: cmd/etcd.go (hunk 16 above)
Comment (xiekeyi98) [reply cmd/etcd.go:102]: If we do not check it, indeed, It will not delete all the keys.

Comment (xiekeyi98) [reply cmd/etcd.go:102]: done.

File: cmd/etcd.go (hunk 17 above)
Comment (xiekeyi98) [reply cmd/etcd.go:148]: done.

**Review by winkyao — APPROVED** [review]
LGTM

**Pull request merged**